### ISSUE TITLE
perf: per-rank batch size hardcoded to 1 — RCA

### ISSUE THREAD

**Issue by RohanAwhad** [RohanAwhad/rohans-sdft#21]
## Current Behavior

Per-rank micro-batch is **hardcoded to 1** — one sample per student forward pass, on every rank. Effective batch size is simulated *only* via `GRAD_ACCUM_STEPS` (which is itself split across ranks: `local_accum_steps = GRAD_ACCUM_STEPS // world_size`, with `assert GRAD_ACCUM_STEPS % world_size == 0` at trainer.py:637).

**RCA — the batch-1 chain:**

1. `megatron_trainer/config.py:78` — hardcoded, **not env-overridable** (unlike `GRAD_ACCUM_STEPS`):

```python
BATCH_SIZE = 1  # always 1; effective batch = BATCH_SIZE * GRAD_ACCUM_STEPS
```

2. `megatron_trainer/trainer.py:715` — every DataLoader "item" is exactly one sample:

```python
dataloader = DataLoader(dataset, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collator, drop_last=True)
```

3. `megatron_trainer/trainer.py:923, 937, 949-953` — rank 0 pulls `GRAD_ACCUM_STEPS` items per step, slices `local_accum_steps` per rank, then calls `_train_sample` **once per item** (one forward per sample):

```python
items = [next(data_iter) for _ in range(GRAD_ACCUM_STEPS)]      # rank 0
my_items = rollout_data[rank * local_accum_steps : (rank + 1) * local_accum_steps]
for micro_step, item_data in enumerate(my_items):
    result = _train_sample(item_data, micro_step, local_accum_steps, ...)
```

4. `megatron_trainer/trainer.py:374, 377` — batch dim is forced to 1 by construction:

```python
input_ids = torch.cat([prompt_ids, comp_ids_t]).unsqueeze(0)   # (1, S) — batch = 1
position_ids = torch.arange(input_ids.size(1), ...).unsqueeze(0)
```

5. Teacher logprob server is batch-1 too: `megatron_trainer/logprob_server.py:186-187` (`ids.unsqueeze(0)`), one TCP request per sample.

6. Legacy guard makes it explicit in the old trainer: `train_dir/src/trainer.py:328`:

```python
if BATCH_SIZE != 1: raise ValueError(f'ONLY batch size == 1 is supported. we got {BATCH_SIZE}')
```

**Impact:** per-forward GEMMs are single-sequence → kernel-launch bound for short completions, poor GPU utilization; throughput scales only via `GRAD_ACCUM_STEPS` (more optimizer steps, no per-forward efficiency gain).

## Desired Behavior

Configurable per-rank micro-batch (>1) so the student forward processes N samples per call, raising GEMM utilization and wall-clock throughput without touching the effective-batch semantics (grad accum / IS ratios unchanged).

## Why it was 1

Memory: at seq 8192 (MAX_TOTAL_LEN) a batch-2 forward on the 8B (or 20B FSDP) risks OOM — the choice was deliberate (batch-1 per forward + accum), but the hardcode is now the constraint. See follow-up comment for ranked options.

**Comment by RohanAwhad** [comment]
## Solutions (ranked)

### 1 (Recommended) — Sequence packing: N samples per forward, batch dim stays 1

Pack `N` samples back-to-back (prompt+completion, separator between) into ONE sequence; forward token count × N, so GEMMs get bigger without any padding waste. This is the verl pattern (verl packs multiple samples per sequence with block-diagonal attention).

- **Where:** collator (`megatron_trainer/collator.py`) packs items; `_train_sample` consumes a packed sequence with per-segment boundaries; `make_kl_processor` + `chunked_head` compute the reverse-KL per segment (flash-attn block-diagonal mask, or segment-id filtering); teacher logprobs stay per-sample (unchanged TCP path).
- **Pros:** best utilization/memory ratio — no padding overhead, activation memory grows linearly with real tokens only; proven upstream (verl GRPO/PPO).
- **Cons:** biggest code change (segment-aware attention + KL processor); needs a flash-attn or block-diagonal mask path in MCore; seq length per forward must stay under `MAX_TOTAL_LEN` budget.

### 2 — True micro-batching: `BATCH_SIZE` becomes env-overridable, batched forward

Make `BATCH_SIZE` env-configurable (like `GRAD_ACCUM_STEPS`), `DataLoader` yields batches of N, `_train_sample` → `_train_batch` with padded `[N, S]` inputs + per-row position ids + attention mask; drop the legacy `BATCH_SIZE != 1` guard.

- **Where:** `config.py:78` (env var), `trainer.py:715/949-953` (batch loop), `_train_sample` (list → batched tensors), KL processor (batched logits), `logprob_server.py` (optional batched teacher request to avoid N TCP round trips).
- **Pros:** simplest change; immediate GEMM-util win; keeps effective-batch semantics (accum unchanged).
- **Cons:** padding waste at variable completion lengths (completions vary ~100-6000 tokens → up to 50%+ waste); OOM risk at seq 8192 — batch-2 on the 8B may already exceed memory, which is likely why it was 1. Needs a memory check first.

### 3 — Stopgap: keep forward batch=1, batch everything around it

No student-side change. Gains come from server-side batching already in place: vLLM `max_num_seqs` batches the 32 concurrent rollout requests internally (per-env requests are already concurrent via `ThreadPoolExecutor`); teacher logprob server gets a batched request path (one forward for N samples instead of N forwards); raise `GRAD_ACCUM_STEPS` for bigger effective batches.

- **Where:** `vllm_utils.py` (nothing), `logprob_server.py` (batch dim in `compute_logprobs_fp16`, batched TCP protocol), `train_full.sh` (env knobs).
- **Pros:** zero OOM risk, small diff, immediate rollout/teacher throughput gains.
- **Cons:** leaves the student forward batch-1 → the per-forward GEMM utilization problem remains (the actual bottleneck for short completions).

**Verdict:** do #2 as the quick win only if memory allows at your seq/rank config; invest in #1 when the 20B scale demands it; #3 in parallel regardless (teacher batching is independent and low-risk).
